File: src/main/java/com/github/sorusclient/client/setting/Setting.java
package com.github.sorusclient.client.setting;

import java.util.List;

public class Setting<T> {

    private final Class<T> type;
    private T value;
    private List<T> forcedValues;

    @SuppressWarnings("unchecked")
    public Setting(T defaultValue) {
        this((Class<T>) defaultValue.getClass(), defaultValue);
    }

    public <U extends T> Setting(Class<T> clazz, U defaultValue) {
        this.type = clazz;
        this.value = defaultValue;
    }

    public Class<T> getType() {
        return this.type;
    }

    public void setForcedValueRaw(List<Object> forcedValues) {
        this.forcedValues = (List<T>) forcedValues;
    }

    public boolean isForcedValue() {
        return this.forcedValues != null;
    }

    public List<T> getForcedValues() {
        return forcedValues;
    }

    public void setValue(T value) {
        this.value = value;
    }

    @SuppressWarnings("unchecked")
    public void setValueRaw(Object value) {
        this.value = (T) value;
    }

    public T getValue() {
        if (this.forcedValues != null) {
            return this.forcedValues.contains(this.value) ? this.value : this.forcedValues.get(0);
        } else {
            return this.value;
        }
    }

    public T getRealValue() {
        return this.value;
    }

}

File: src/main/kotlin/com/github/sorusclient/client/setting/Setting.kt
package com.github.sorusclient.client.setting

class Setting<T>(val type: Class<T>, val defaultValue: T) {

    var realValue: T
    var forcedValues: List<T>? = null
        private set

    private var parentValue: T? = null

    var overriden = true

    val value: T
        get() {
            val realValue = if (overriden || parentValue == null) { realValue } else { parentValue!! }
            return if (forcedValues != null) {
                if (forcedValues!!.contains(realValue)) realValue else forcedValues!![0]
            } else {
                realValue
            }
        }

    constructor(defaultValue: T) : this(defaultValue!!::class.java as Class<T>, defaultValue)

    init {
        realValue = defaultValue
    }

    fun setForcedValueRaw(forcedValues: List<Any>?) {
        this.forcedValues = forcedValues as List<T>?
    }

    val isForcedValue: Boolean
    get() = forcedValues != null

    fun setValueRaw(value: Any) {
        realValue = value as T
    }

    fun setValueRaw(value: Any, isPrimary: Boolean) {
        if (!isPrimary) {
            parentValue = value as T
        }
        overriden = isPrimary
        realValue = value as T
    }

}